src/inmanta/agent/handler.py

```diff
@@ -419,31 +420,50 @@ def _log_msg(self, level: int, msg: str, *args: object, exc_info: bool = False,
 
 
 @stable_api
-class ResourceHandler(object):
+class HandlerABC(ABC, Generic[R]):
     """
-    A baseclass for classes that handle resources. New handler are registered with the
-    :func:`~inmanta.agent.handler.provider` decorator.
-
-    The implementation of a handler should use the ``self._io`` instance to execute io operations. This io objects
-    makes abstraction of local or remote operations. See :class:`~inmanta.agent.io.local.LocalIO` for the available
-    operations.
+    Top-level abstract base class all handlers should inherit from. This class
+    is generic with regard to the resource type this handler is responsible for.
 
-    :param agent: The agent that is executing this handler.
-    :param io: The io object to use.
+    :param agent: The agent responsible for this handler
     """
 
-    def __init__(self, agent: "inmanta.agent.agent.AgentInstance", io: Optional["IOBase"] = None) -> None:
+    def __init__(self, agent: "inmanta.agent.agent.AgentInstance") -> None:
         self._agent = agent
 
-        if io is None:
-            raise Exception("Unsupported: no resource mgmt in RH")
-        else:
-            self._io = io
-
         self._client: Optional[protocol.SessionClient] = None
         # explicit ioloop reference, as we don't want the ioloop for the current thread, but the one for the agent
         self._ioloop = agent.process._io_loop
 
+    def pre(self, ctx: HandlerContext, resource: R) -> None:
+        """
+        Method executed before a handler operation (Facts, dryrun, real deployment, ...) is executed. Override this method
+        to run before an operation.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The resource to query facts for.
+        """
+
+    def post(self, ctx: HandlerContext, resource: R) -> None:
+        """
+        Method executed after an operation. Override this method to run after an operation.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The resource to query facts for.
+        """
+
+    @abstractmethod
+    def deploy(
+        self,
+        ctx: HandlerContext,
+        resource: R,
+        requires: Dict[ResourceIdStr, ResourceState],
```

Comment:
I think this would be a good time to clean up this type. Either `dict` or `abc.Mapping`. Afaik we don't ever update it so `Mapping` would be best. @arnaudsjs do you know?

Reply:
I don't think we ever update it.